### ISSUE TITLE
Implementation details for Exchange and Illness datasets

### ISSUE THREAD

**Issue by NamburiSrinath** [ACAT-SCUT/CycleNet#7]
Hi @lss-1138 and @wayhoww 

Impressive work, congrats!!

I am wondering how to use this repo on Exchange and Illness datasets as there are no scripts for those datasets. 

Below is an example script I've given a try

```
model_name=CycleNet

root_path_name=./dataset/
data_path_name=exchange_rate.csv
model_id_name=exchange
data_name=custom

model_type='linear'
seq_len=96
pred_len=48
for random_seed in 2024 2025 2026 2027 2028
do
    python -u run.py \
      --is_training 1 \
      --root_path $root_path_name \
      --data_path $data_path_name \
      --model_id $model_id_name'_'$seq_len'_'$pred_len \
      --model $model_name \
      --data $data_name \
      --features M \
      --seq_len $seq_len \
      --pred_len $pred_len \
      --enc_in 21 \
      --cycle 144 \
      --model_type $model_type \
      --train_epochs 30 \
      --patience 5 \
      --itr 1 --batch_size 256 --learning_rate 0.01 --random_seed $random_seed > $model_id_name'_'$seq_len'_'$pred_len'_'$random_seed.txt
done
```

and this is the error

```
Traceback (most recent call last):                                             
  File "/hdd4/srinath2/CycleNet/run.py", line 145, in <module>
    exp.train(setting)                 
  File "/hdd4/srinath2/CycleNet/exp/exp_main.py", line 182, in train
    outputs = self.model(batch_x, batch_cycle)                                 
  File "/hdd4/srinath2/.conda/envs/core/lib/python3.10/site-packages/torch/nn/m
odules/module.py", line 1553, in _wrapped_call_impl                            
    return self._call_impl(*args, **kwargs)                                    
  File "/hdd4/srinath2/.conda/envs/core/lib/python3.10/site-packages/torch/nn/m
odules/module.py", line 1562, in _call_impl                                    
    return forward_call(*args, **kwargs)                                       
  File "/hdd4/srinath2/CycleNet/models/CycleNet.py", line 55, in forward
    x = x - self.cycleQueue(cycle_index, self.seq_len)                         
RuntimeError: The size of tensor a (8) must match the size of tensor b (21) at 
non-singleton dimension 2
```

 Any help in understanding settings (`enc_in, cycles`) will be super helpful.

**Comment by lss-1138** [comment]
Thank you for your attention.

1. **`enc_in`** refers to the number of variables (or channels) in the dataset. For the Exchange dataset, it should be set to 8, and for the ILI dataset, it should be set to 7. If the `enc_in` setting does not match the actual number of variables in the dataset, the program will throw an error like the one you encountered.

2. **`cycle`** refers to the intrinsic periodic cycle length in the dataset. As shown in Table 1 and Figure 6 of our paper, the cycle length can be determined by analyzing the periodic patterns of the dataset or using tools like the ACF (Autocorrelation Function). 

However, the Exchange dataset does not have a stable periodic cycle. This is because financial data is inherently unpredictable and lacks consistent periodicity. Additionally, by analyzing the ACF plot, we can see that the Exchange data does not exhibit a strong periodic pattern. In this case, choosing an appropriate hyperparameter for the cycle length (W) is challenging. If you must set it, we suggest using **W=7**, but please note that CycleNet may not perform optimally here, as it is designed to model periodic patterns in data.

For the ILI dataset, the cycle is yearly (with weekly sampling). As discussed in the **Discussion** section of our paper, CycleNet has limitations when modeling long-range cycles. The model may not have enough data to fully capture the periodic pattern, and as a result, CycleNet might not perform at its best in such scenarios.

In summary, if you wish to try these datasets, we recommend the following settings, though CycleNet might not perform at its best in such scenarios:
- **Exchange**: `enc_in=8`, `cycle=7`
- **Illness**: `enc_in=7`, `cycle=52`

I hope this helps!

**Comment by NamburiSrinath** [comment]
Thanks a ton for the detailed comments. This was super helpful!

I've some general follow-up questions. For a new dataset

- `enc_in`: An easy way to know what's the `enc_in` for a new dataset is to just run the code, look at the error message and adjust it.
- `cycle`: How can I get the cycle for a new dataset? The ACF plot suggests whether there's a periodicity and if there's one such, that'll be the cycles

**Question**: Is the code for ACF plots (The figure 6) in this repo? Also, is it because the **W** is set to cycle length that the peaks are observed at lag 0?

**Comment by lss-1138** [comment]
We’re glad to hear that you found the information helpful!  

We’ve just added **[acf_plot.ipynb](https://github.com/ACAT-SCUT/CycleNet/blob/main/acf_plot.ipynb)** to the repository, which is the original notebook used to generate the results in Figure 6.  

Regarding your question—whether the peaks in the ACF plot appear at lag 0 because **W** is set to the cycle length—the answer is no. The peaks in the ACF plot are determined solely by the data's inherent structure and periodicity, not by the **W** parameter used in the model.  

We believe that reviewing the workflow in **[acf_plot.ipynb](https://github.com/ACAT-SCUT/CycleNet/blob/main/acf_plot.ipynb)** will help clarify this and enhance your understanding.

**Comment by NamburiSrinath** [comment]
Got it and totally makes sense. 

The caption in fig 6 - `The hyperparameter W should be set to the lag corresponding to the observed maximum peak.`
I believe this is to see for what value of lag, the autocorrelation becomes that of lag 0 (which has a peak).

Once again, thanks for the prompt response.
